ocrd_network/ocrd_network/processing_server.py

```diff
@@ -0,0 +1,289 @@
+
+from ocrd_utils import (
+    getLogger,
+    get_ocrd_tool_json,
+)
+from ocrd_validators import (
+    ParameterValidator,
+    ProcessingServerValidator
+)
+from ocrd_network.database import initiate_database
+from ocrd_network.deployer import Deployer
+from ocrd_network.deployment_config import ProcessingServerConfig
+from ocrd_network.rabbitmq_utils import (
+    RMQPublisher,
+    OcrdProcessingMessage
+)
+from ocrd_network.models.job import (
+    Job,
+    JobInput,
+    JobOutput,
+    StateEnum
+)
+from ocrd_network.models.workspace import Workspace
+
+
+class ProcessingServer(FastAPI):
+    """FastAPI app to make ocr-d processor calls
+
+    The Processing-Server receives calls conforming to the ocr-d webapi regarding the processing
+    part. It can run ocrd-processors and provides endpoints to discover processors and watch the job
+    status.
+    The Processing-Server does not execute the processors itself but starts up a queue and a
+    database to delegate the calls to processing workers. They are started by the Processing-Server
+    and the communication goes through the queue.
+    """
+
+    def __init__(self, config_path: str, host: str, port: int) -> None:
+        super().__init__(on_startup=[self.on_startup], on_shutdown=[self.on_shutdown],
+                         title="OCR-D Processing Server",
+                         description="OCR-D processing and processors")
+        self.log = getLogger(__name__)
+        self.hostname = host
+        self.port = port
+        self.config = ProcessingServer.parse_config(config_path)
+        self.deployer = Deployer(self.config)
+        self.mongodb_url = None
+        self.rmq_host = self.config.queue.address
+        self.rmq_port = self.config.queue.port
+        self.rmq_vhost = '/'
+        self.rmq_username = self.config.queue.credentials[0]
+        self.rmq_password = self.config.queue.credentials[1]
+
+        # Gets assigned when `connect_publisher` is called on the working object
+        self.rmq_publisher = None
+
+        # This list holds all processors mentioned in the config file
+        self._processor_list = None
+
+        # Create routes
+        self.router.add_api_route(
+            path='/stop',
+            endpoint=self.stop_deployed_agents,
+            methods=['POST'],
+            tags=['tools'],
+            summary='Stop database, queue and processing-workers',
+        )
+
+        self.router.add_api_route(
+            path='/processor/{processor_name}',
+            endpoint=self.push_processor_job,
+            methods=['POST'],
+            tags=['processing'],
+            status_code=status.HTTP_200_OK,
+            summary='Submit a job to this processor',
+            response_model=JobOutput,
+            response_model_exclude_unset=True,
+            response_model_exclude_none=True
+        )
+
+        self.router.add_api_route(
+            path='/processor/{_processor_name}/{job_id}',
```

Comment:
`path` can just be `/processor/{job_id}`. `processor_name` is not needed here.

Reply:
@joschrew, is there a specific reason for this or can this be refactored?

Reply:
@tdoan2010, while checking the [spec](https://github.com/OCR-D/spec/blob/master/web_api.md) (to implement the callback_url), I saw that this is still the path specified in the spec, hence, the path here is `/processor/{executable}/{job_id}`.

Reply:
OK, I see the difference now. How does the MongoDB look like in this implementation? I suppose it has only 1 collection named `Job`, right?

Reply:
Right, there's just `Job` for now. The current version of the WebAPI implementation has just `WorkflowJob`.

Reply:
> @tdoan2010, while checking the [spec](https://github.com/OCR-D/spec/blob/master/web_api.md) (to implement the callback_url), I saw that this is still the path specified in the spec, hence, the path here is `/processor/{executable}/{job_id}`.

That was under the assumption that `job-id` are sequential per executable, i.e. having both `/procesor/foo/1` and `/processor/bar/1`. But if they're UUID, the `{executable}` path variable is not necessary, feel free to change the spec accordingly.

Reply:
Tracked [here](https://github.com/OCR-D/spec/issues/246) – resolving.